plugins/kafka/kafka.cpp

```diff
@@ -0,0 +1,1167 @@
+                                #endif
+                                shouldRead = false;
+                                break;
+
+                            case RdKafka::ERR__UNKNOWN_PARTITION:
+                                // Unknown partition; don't throw an error here because
+                                // in some configurations (e.g. more Thor slaves than
+                                // partitions) not all consumers will have a partition
+                                // to read
+                                #ifdef _DEBUG
+                                    DBGLOG("Kafka: Unknown partition while trying to read");
+                                #endif
+                                shouldRead = false;
+                                break;
+
+                            case RdKafka::ERR__UNKNOWN_TOPIC:
+                                throw MakeStringException(-1, "Kafka: Error while reading message: '%s'", messageObjPtr->errstr().c_str());
+                                break;
+                        }
+                    }
+                    catch (...)
+                    {
+                        delete(messageObjPtr);
+                        throw;
+                    }
+
+                    delete(messageObjPtr);
+                }
+            }
+        }
+
+        return result;
+    }
+
+    void KafkaStreamedDataset::stop()
+    {
+        shouldRead = false;
+    }
+
+    //--------------------------------------------------------------------------
+
+    Poller::Poller(KafkaObj* _parentPtr, __int32 _pollTimeout)
+        :   Thread("Kafka::Poller"),
+            parentPtr(_parentPtr),
+            pollTimeout(_pollTimeout)
+    {
+        shouldRun = false;
+        running = false;
+    }
+
+    Poller::~Poller()
+    {
+
+    }
+
+    void Poller::start()
+    {
+        if (!running)
+        {
+            shouldRun = true;
+            Thread::start();
+        }
+    }
+
+    void Poller::stop()
+    {
+        if (running)
+        {
+            shouldRun = false;
+
+            while (running)
+            {
+                usleep(pollTimeout);
+            }
+        }
+    }
+
+    int Poller::run()
+    {
+        if (parentPtr)
+        {
+            RdKafka::Handle*    handle = parentPtr->handle();
+
+            running = true;
+
+            while (shouldRun)
+            {
+                handle->poll(pollTimeout);
+            }
+        }
+
+        running = false;
+
+        return 0;
+    }
+
+    //--------------------------------------------------------------------------
+
+    Publisher::Publisher(const std::string& _brokers, const std::string& _topic, __int32 _pollTimeout)
+        :   brokers(_brokers),
+            topic(_topic),
+            pollTimeout(_pollTimeout)
+    {
+        producerPtr = NULL;
+        topicPtr = NULL;
+        lockPtr = new Mutex;
```

Comment:
Converted all mutexes to CriticalSection members instead (plus one static variable for initialization).